enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,159 @@
+
+## Summary
+
+This enhancement describes capturing login, login failure, and logout events in the OpenShift authentication subsystems.
+
+Currently, login and logout events can only be derived from inspecting audit logs. These include creation or deletion of `oauthaccesstokens` resources as described in the API server [Audit Policy Configuration](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/audit-policy.md#logging-of-token-with-secure-oauth-storage).
+
+This approach has a couple of drawbacks:
+1. Login events are captured via creation events of `oauthaccesstokens` in audit logs. This includes successful login events only, not login failures.
+2. Logout events are captured via deletion events of `oauthaccesstokens` in audit logs. Here only references to the sha256 based access token are present in audit logs, but no additional information about the underlying username who logged out.
+
+## Motivation
+
+### Goals
+
+1. Add capabilities in oauth-server to create audit-like logs. Gate fine grained audit log events via a simple profile configuration mechanism.
+2. Extend oauth-server to capture login and login failure events.
+3. Extend oauth-apiserver to add user information when deleting `oauthaccesstoken` objects.
+4. Extend must-gather/gather_audit_logs to fetch oauth-server audit logs.
+
+### Non-Goals
+
+1. Hook into the existing API server [policy configuration](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/audit-policy.md).
+2. Provide a fully featured audit subsystem in oauth-server compliant with API server policy framework.
+
+## Proposal
+
+We propose to add an `audit` struct to `oauth.config.openshift.io/v1` with a `profile` field which configures audit logging granularity.
+
+```go
+// OAuthSpec contains desired cluster auth configuration
+type OAuthSpec struct {
+	// identityProviders is an ordered list of ways for a user to identify themselves.
+	// When this list is empty, no identities are provisioned for users.
+	// +optional
+	IdentityProviders []IdentityProvider `json:"identityProviders,omitempty"`
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+
+	Audit OAuthAudit `json:"audit"`
+}
+
+type OAuthAudit struct {
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+  // This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+### As a user I want to inspect login and login failure attempts
```

Comment:
lang nit: `As a user I want to inspect successful and failed logins`
(I think "login failure attempt" might technically make sense when bruteforcing but I think you meant the above instead :slightly_smiling_face: )

Reply:
yes, that's a great suggestion :+1:  fixed